components/constellation/constellation.rs

```diff
@@ -849,28 +849,38 @@ where
                     ipc_scheduler_receiver,
                 );
 
-                let (background_hang_monitor_sender, ipc_bhm_receiver) =
-                    ipc::channel().expect("ipc channel failure");
-                let background_hang_monitor_receiver =
-                    route_ipc_receiver_to_new_mpsc_receiver_preserving_errors(ipc_bhm_receiver);
+                let (background_hang_monitor_sender, background_hang_monitor_receiver) =
+                    if opts::get().background_hang_monitor {
+                        let (bhm_sender, ipc_bhm_receiver) =
+                            ipc::channel().expect("ipc channel failure");
+                        (
+                            Some(bhm_sender),
+                            Some(route_ipc_receiver_to_new_mpsc_receiver_preserving_errors(
+                                ipc_bhm_receiver,
+                            )),
+                        )
+                    } else {
+                        (None, None)
+                    };
 
                 // If we are in multiprocess mode,
                 // a dedicated per-process hang monitor will be initialized later inside the content process.
                 // See run_content_process in servo/lib.rs
-                let (background_monitor_register, sampler_chan) = if opts::multiprocess() {
-                    (None, vec![])
-                } else {
-                    let (sampling_profiler_control, sampling_profiler_port) =
-                        ipc::channel().expect("ipc channel failure");
-
-                    (
-                        Some(HangMonitorRegister::init(
-                            background_hang_monitor_sender.clone(),
-                            sampling_profiler_port,
-                        )),
-                        vec![sampling_profiler_control],
-                    )
-                };
+                let (background_monitor_register, sampler_chan) =
+                    if opts::multiprocess() | !opts::get().background_hang_monitor {
+                        (None, vec![])
+                    } else {
+                        let (sampling_profiler_control, sampling_profiler_port) =
+                            ipc::channel().expect("ipc channel failure");
+
+                        (
+                            Some(HangMonitorRegister::init(
+                                background_hang_monitor_sender.clone().unwrap(),
```

Comment:
To fix the panic lint, you have to remove the `unwrap` here. It should never panic, but it's just a lint on the `unwrap`. So you can return `(None, vec![])` here as well, and `warn`, if `background_hang_monitor_sender` is `None`.

Reply:
If I put a check with `is_some()` and return above if it's true but `unwrap()` would still be used inside the `if` block. Won't that be a problem?

Reply:
So the idea is that there is no `unwrap` in the code, even if logically it will never panic :)

I think you literally do `If let Some(bhm_sender) =`, and put the code in there, then in the `else` branch you `warn!(No BHM sender found in BHM mode.)` and return `(None, vec![])`

Reply:
Ah, my bad. I completely forgot about the `if let` pattern matching. I was entirely relying on the methods defined on Option enum.